### DISCUSSION TITLE
使用Mingw cannot get program for cu

### DISCUSSION THREAD

**Discussion by chaseaishang** [xmake-io/xmake#3524] [Q&A] [answered]
使用MSVS可以正常编译cuda.
但是切换为其他就报错,cannot get program for cu

PS <> xmake -D -v
error: @programdir\core\main.lua:302: @programdir\actions\build\main.lua:206: ...mdir\core\sandbox\modules\import\core\base\scheduler.lua:96: ...mdir\core\sandbox\modules\import\core\base\scheduler.lua:56: @programdir\modules\private\async\runjobs.lua:256: ...amdir\core\sandbox\modules\import\core\tool\compiler.lua:37: cannot get program for cu
stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:913]:
    [...amdir\core\sandbox\modules\import\core\tool\compiler.lua:37]: in function 'load'
    [@programdir\modules\private\action\build\object.lua:40]: in function 'script'
    [@programdir\modules\private\action\build\object.lua:104]: in function 'build_object'
    [@programdir\modules\private\action\build\object.lua:129]: in function 'jobfunc'
    [@programdir\modules\private\async\runjobs.lua:232]:

stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:913: in function 'base/os.raiselevel'
        (...tail calls...)
        @programdir\modules\private\async\runjobs.lua:256: in field 'catch'
        @programdir\core\sandbox\modules\try.lua:123: in global 'try'
        @programdir\modules\private\async\runjobs.lua:218: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
stack traceback:
        @programdir\core\base\utils.lua:290: in function <@programdir\core\base\utils.lua:280>
        [C]: in function 'error'
        @programdir\core\base\os.lua:913: in function 'base/os.raiselevel'
        (...tail calls...)
        ...mdir\core\sandbox\modules\import\core\base\scheduler.lua:56: in field 'co_start_withopt'
        @programdir\modules\private\async\runjobs.lua:217: in function <@programdir\modules\private\async\runjobs.lua:162>
        [C]: in function 'xpcall'
        @programdir\core\base\utils.lua:280: in function 'base/utils.trycall'
        (...tail calls...)
        @programdir\core\base\task.lua:501: in function 'base/task.run'
        @programdir\core\main.lua:300: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:913: in function 'base/os.raiselevel'
        (...tail calls...)
        @programdir\core\main.lua:302: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
        @programdir\core\base\utils.lua:290: in function <@programdir\core\base\utils.lua:280>
        [C]: in function 'error'
        @programdir\core\base\os.lua:913: in function 'os.raiselevel'
        (...tail calls...)
        ...mdir\core\sandbox\modules\import\core\base\scheduler.lua:56: in field 'co_start_withopt'
        @programdir\modules\private\async\runjobs.lua:217: in function <@programdir\modules\private\async\runjobs.lua:162>
        [C]: in function 'xpcall'
        @programdir\core\base\utils.lua:280: in function 'base/utils.trycall'
        @programdir\core\base\scheduler.lua:520: in function 'base/scheduler.co_group_begin'
        ...mdir\core\sandbox\modules\import\core\base\scheduler.lua:94: in field 'co_group_begin'
        @programdir\modules\private\async\runjobs.lua:162: in function <@programdir\modules\private\async\runjobs.lua:56>
        (...tail calls...)
        @programdir\actions\build\build.lua:259: in function <@programdir\actions\build\build.lua:247>
        (...tail calls...)
        @programdir\actions\build\main.lua:109: in global '_do_build'
        @programdir\actions\build\main.lua:183: in function <@programdir\actions\build\main.lua:173>
        [C]: in function 'xpcall'
        @programdir\core\base\utils.lua:280: in function 'base/utils.trycall'
        @programdir\core\sandbox\modules\try.lua:117: in global 'try'
        @programdir\actions\build\main.lua:171: in function <@programdir\actions\build\main.lua:133>
        (...tail calls...)
        [C]: in function 'xpcall'
        @programdir\core\base\utils.lua:280: in function 'base/utils.trycall'
        (...tail calls...)
        @programdir\core\base\task.lua:501: in function 'base/task.run'
        @programdir\core\main.lua:300: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>

**Comment by waruqi** [comment]
你怎么切的。。这边没对 mingw cuda 做过特别支持。

**Reply by chaseaishang** [reply]
就用这个xmake f -c -p mingw -vD

**Reply by chaseaishang** [reply]
看着都找到了。。。。。
checking for architecture ... x86_64
checking for mingw directory ... E:\DevelopEnvironment\MinGW\mingw64
checking for Cuda SDK directory ... C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0
checking for C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0\bin\nvcc ... ok
checking for cuda devices
> DEVICE #0
>     name = "NVIDIA GeForce RTX 2060"
>     totalGlobalMem = 6442123264
>     sharedMemPerBlock = 49152
>     regsPerBlock = 65536
>     warpSize = 32
>     memPitch = 2147483647
>     maxThreadsPerBlock = 1024
>     maxThreadsDim = (1024, 1024, 64)
>     maxGridSize = (2147483647, 65535, 65535)
>     clockRate = 1680000
>     totalConstMem = 65536
>     major = 7
>     minor = 5
>     textureAlignment = 512
>     texturePitchAlignment = 32
>     deviceOverlap = true
>     multiProcessorCount = 30
>     kernelExecTimeoutEnabled = false
>     integrated = false
>     canMapHostMemory = true
>     computeMode = 0
>     maxTexture1D = 131072
>     maxTexture1DMipmap = 32768
>     maxTexture1DLinear = 268435456
>     maxTexture2D = (131072, 65536)
>     maxTexture2DMipmap = (32768, 32768)
>     maxTexture2DLinear = (131072, 65000, 2097120)
>     maxTexture2DGather = (32768, 32768)
>     maxTexture3D = (16384, 16384, 16384)
>     maxTexture3DAlt = (8192, 8192, 32768)
>     maxTextureCubemap = 32768
>     maxTexture1DLayered = (32768, 2048)
>     maxTexture2DLayered = (32768, 32768, 2048)
>     maxTextureCubemapLayered = (32768, 2046)
>     maxSurface1D = 32768
>     maxSurface2D = (131072, 65536)
>     maxSurface3D = (16384, 16384, 16384)
>     maxSurface1DLayered = (32768, 2048)
>     maxSurface2DLayered = (32768, 32768, 2048)
>     maxSurfaceCubemap = 32768
>     pciDomainID = 0
>     tccDriver = false
>     asyncEngineCount = 2
>     unifiedAddressing = true
>     memoryClockRate = 7001000
>     memoryBusWidth = 192
>     l2CacheSize = 3145728
>     maxThreadsPerMultiProcessor = 1024
>     streamPrioritiesSupported = true
>     globalL1CacheSupported = true
>     localL1CacheSupported = true
>     sharedMemPerMultiprocessor = 65536
>     regsPerMultiprocessor = 65536
>     isMultiGpuBoard = false
>     multiGpuBoardGroupID = 0
>     singleToDoublePrecisionPerfRatio = 32
>     pageableMemoryAccess = false
>     concurrentManagedAccess = false
>     managedMemory = true
>     computePreemptionSupported = true
>     canUseHostPointerForRegisteredMem = false
>     cooperativeLaunch = true
>     cooperativeMultiDeviceLaunch = false
>     sharedMemPerBlockOptin = 65536
>     pageableMemoryAccessUsesHostPageTables = false
>     directManagedMemAccessFromHost = false
>     uuid = "c8b1cd5e-2c0f-fa19-2e64-c7d89f438936"
>     luid = "35fd000000000000"
>     luidDeviceNodeMask = 1
> found device #0: NVIDIA GeForce RTX 2060 with compute 7.5 capability
configure
{
    arch = x86_64
    ccache = true
    ndk_stdcxx = true
    buildir = build
    cuda = C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0
    host = windows
    mingw = E:\DevelopEnvironment\MinGW\mingw64
    clean = true
    mode = release
    plat = mingw
    kind = static
}

**Reply by chaseaishang** [reply]
换成clang也不行
PS E:\workplace\c++\testxmake\test> xmake
checking for Microsoft Visual Studio (x64) version ... 2022
checking for Cuda SDK directory ... C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0
[ 25%]: compiling.release src\main.cu
error: clang: error: unknown argument: '-nologo'
clang: error: unknown argument: '-EHsc'
clang: error: no such file or directory: 'cuda_runtime.h'
clang: error: no such file or directory: 'E:workplacec++testxmaketestsrcmain.cu'
clang: error: no input files
  > in src\main.cu
warning: add_cuflags("-rdc=true") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
PS E:\workplace\c++\testxmake\test>

**Comment by waruqi** [comment, marked as answer]
目前 mingw 平台，没自动绑定 cuda 工具链，只有 windows 平台才绑定了。。

就算 mingw 下，调用 nvcc 编译 .cu，默认应该也是用的 msvc 的 ccbin，确定跟 mingw c++ 混编兼容么？

另外，我只知道 windows 上 nvcc 支持 msvc 的 ccbin ，其他 clang mingw gcc 的 ccbin 也支持？我怎么记得之前 issues 有人说不支持来着。

**Reply by chaseaishang** [reply]
行吧，我找到的也都说Windows下MinGW不行。都是用MSVC
但是我的clang为啥也不行。。。
用CMake也是只有MSVC行。。。
PS E:\workplace\c++\testxmake\test> xmake -rv
[ 25%]: compiling.release src\main.cu
"C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v12.0\\bin\\nvcc" -c -ccbin=clang "-IC:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v12.0\\include" -m64 -gencode arch=compute_75,code=sm_75 -o build\.objs\test\windows\x64\release\src\main.cu.obj src\main.cu
error: clang: error: unknown argument: '-nologo'
clang: error: unknown argument: '-EHsc'
clang: error: no such file or directory: 'cuda_runtime.h'
clang: error: no such file or directory: 'E:workplacec++testxmaketestsrcmain.cu'
clang: error: no input files
warning: add_cuflags("-rdc=true") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
PS E:\workplace\c++\testxmake\test> xmake -rv
checking for C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0\bin\nvcc ... ok
checking for the cuda compiler (cu) ... nvcc
checking for cl.exe ... E:\vistualStudio\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64\cl.exe
checking for the c compiler (cc) ... cl.exe
checking for C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0\bin\C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.0\bin\nvcc ... no
checking for flags (-O3) ... no
checking for flags (-rdc=true) ... no
checking for flags (-gencode arch=compute_75,code=sm_75) ... ok
[ 25%]: compiling.release src\main.cu
"C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v12.0\\bin\\nvcc" -c -ccbin=clang "-IC:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v12.0\\include" -m64 -gencode arch=compute_75,code=sm_75 -o build\.objs\test\windows\x64\release\src\main.cu.obj src\main.cu
checking for flags (-MD -MF) ... no
checking for flags (-M) ... no
error: clang: error: unknown argument: '-nologo'
clang: error: unknown argument: '-EHsc'
clang: error: no such file or directory: 'cuda_runtime.h'
clang: error: no such file or directory: 'E:workplacec++testxmaketestsrcmain.cu'
clang: error: no input fileswarning: .\xmake.lua:6: cannot match add_files("*.cu") in target(test)
warning: add_cuflags("-rdc=true") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
PS E:\workplace\c++\testxmake\test>

**Reply by waruqi** [reply]
> 但是我的clang为啥也不行。。。

我记得之前有人 issues 反馈过，windows 上 ccbin 仅仅支持 cl.exe 。。你切 clang 也不支持。。

只有 linux 下，你可以切 clang

**Reply by waruqi** [reply]
这跟 xmake 好像也没啥关系。。你可以直接调 nvcc 命令试，不行，那就是 nvcc 的问题。。你得到 nvcc 那边反馈

**Reply by chaseaishang** [reply]
ok,ok。看来还是不支持。我之前不知道在哪看到xmake可以支持其他cxx编译器编译。。。
我还是老实用MSVC吧。谢谢你了
